pilz_robot_programming/src/pilz_robot_programming/commands.py

```diff
@@ -217,40 +219,59 @@ def _cmd_to_request(self, robot):
         if self._goal is None:
             raise NameError("Goal is not given.")
 
-        goal_constraints = Constraints()
-
-        # goal as Pose in Cartesian space
-        if isinstance(self._goal, Pose):
-            goal_pose = self._get_goal_pose(robot)
-
-            robot_reference_frame = robot._robot_commander.get_planning_frame()
-            goal_constraints.orientation_constraints.append(
-                _to_ori_constraint(goal_pose, robot_reference_frame, self._target_link))
-            goal_constraints.position_constraints.append(
-                _to_pose_constraint(goal_pose, robot_reference_frame, self._target_link))
-
-        # goal as list of int or float in joint space
-        elif isinstance(self._goal, list):
-            joint_names = robot._robot_commander.get_group(self._planning_group).get_active_joints()
-            joint_values = self._get_joint_pose(robot)
-
-            if len(joint_names) != len(joint_values):
-                raise IndexError("Given joint goal does not match the planning group " + req.group_name + ".")
+        self._check_for_future_goal_expectation()
 
-            for joint_name, joint_value in zip(joint_names, joint_values):
-                joint_constraint = JointConstraint()
-                joint_constraint.joint_name = joint_name
-                joint_constraint.position = joint_value
-                joint_constraint.weight = 1
-                goal_constraints.joint_constraints.append(joint_constraint)
+        convertion_methods = [self._constraint_by_joint_values,
+                              self._constraint_by_pose,
+                              self._constraint_by_stamped_pose]
 
+        for method in convertion_methods:
```

Comment:
fancy, I like that, but with the newer versions python 3 is required, do we really want to use a module, that no longer is supported for python 2?

Reply:
similar to paradigm "duck-typing": Try, if it can convert - otherwise fallback to next function looks very pythonic for me...

Reply:
Yes it could be indeed that coming from C++, this duck-typing and exception throwing is quite suprising to me.
Yet in this case it seems to me that this is also an issue for extensibility and having a simple map seems more straight forward. @martiniil I would like to hear your opinion.

Reply:
I am fine with it if you can argue the following:
- Does not catch too much
- Is not much slower than "asking for permission"

Reply:
Yep, it just catches wrong types during conversion. I've added verbose print to display the problem to the user if conversion fails for all methods.

I'd neglect performance in this case, because it's one conversion, not thousands of commands. Even if we loose [some microseconds](https://stackoverflow.com/questions/2522005/cost-of-exception-handlers-in-python), it's not worth worrying here. But if you insist, I can still timeit to get actual numbers.

Reply:
I am not worried about the exception handlers, but:
From looking at the code I deduce that `_joint_values_to_constraint`only throws after two `robot_commander`-calls. What do you think about the delay from that?

Reply:
`self._robot...` will be refactored in follow-up PR for #251.

I tend to read the start state in the conversion method as soon as possible and make it a member of `BaseCmd` to drop the robot reference... But I'll propose a follow-up PR, if you're okay with that.

Reply:
You could also avoid unnecessary computation by checking for  `hasattr(goal, 'position')` at the beginning of `_constraint_by_pose()`and change the order of the convertion methods to:
```
        convertion_methods = [self._constraint_by_pose,
                              self._constraint_by_stamped_pose,
                              self._constraint_by_joint_values]
```

Reply:
reordered.

(Without hasattr(), see duck-typing and EAFP above)

Reply:
@rfeistenauer explained that one reasoning behind this method is that something convertable by any of the three methods should also be convertable. With this aspect I would accept this way.

Reply:
I am not convinced that duck-typing works here:
```
    def _joint_values_to_constraint(self, joint_names=()):
        if isinstance(self._goal, str):
            raise TypeError("String is not convertible into joint values.")
```